components/omega/doc/design/Logging.md

```diff
@@ -0,0 +1,109 @@
+
+It is a requirement for the OMEGA logging system to furnish Omega sub-components with a mechanism to store log messages onto a file system, based on their corresponding severity levels.
+
+### 2.2 Requirement: Inter-operability with E3SM
+
+It is a requirement for the OMEGA logging system to recognize the log root filename defined in E3SM as a part of the input configuration.
+
+### 2.3 Requirement: Multi-processing
+
+It is a requirement that each process should have the capability to generate log messages independently. Additionally, there should be a mechanism in place to control the number of processes generating log messages simultaneously. This functionality is desired to ensure efficient management of system logs and prevent potential system overloading due to excessive logging activities.
+
+### 2.4 Requirement: Multi-threading
+
+It is a requirement for the OMEGA logging system to provide thread-safe logging capabilities. This functionality is necessary to ensure that concurrent logging activities by multiple threads are executed in a synchronized manner, without causing data corruption or inconsistencies in the generated log files.
+
+### 2.5 Requirement: Omega data types
+
+It is a requirement for the OMEGA logging system to support a subset of Omega data types, in order to generate well-formatted log messages that include these data types. This functionality is crucial to ensure that all relevant information is captured in the logs and to facilitate effective debugging and troubleshooting activities.
+
+### 2.6 Requirement: Buffered file input/output
+
+It is a requirement for the OMEGA logging system to be equipped with buffered file IO capabilities, with the aim of reducing the number of time-consuming file IO operations. The API should provide an option that allows the user to choose between buffered and unbuffered file IO modes. This functionality is crucial to enhance the performance and efficiency of the logging system, by minimizing the overhead associated with file IO operations.
+
+### 2.7 Desired: Formatter
+
+It is desired for the OMEGA logging system to support various use-cases of log files by generating log messages through one of the installed message formatters. These formatters should be switchable at both compile-time and runtime to accommodate changing requirements. The default formatter should generate an unstructured text output, while other possible formatters should be able to generate structured output such as JSON. This functionality is essential to ensure flexibility in the logging system, allowing users to tailor the output format to their specific needs.
+
+### 2.8 Desired: Output sinks
+
+It is desired for the OMEGA logging system to be capable of transporting log messages to various output sinks such as emails or databases. This functionality is crucial for ensuring the accessibility and availability of log data, enabling users to view and analyze the logs through multiple channels.
```

Comment:
Just curious, what do you have in mind regarding output sinks and databases? In the commercial world, they take log analysis seriously using Splunk etc. Of course, we don't need to go all that way.

Reply:
I had a couple of use cases in mind when I put this item:
* Online-training a ML/AL model while Omega model is running
* Realtime visualization of various purposes including animation, debugging, or profiling while model is running
* Other uses cases for data exchange by generating data in well-known structured data format such as JSON.